package/MDAnalysis/lib/mdamath.py

```diff
@@ -341,92 +329,71 @@ def make_whole(atomgroup, reference_atom=None):
-                             "Perhaps the atomgroup isn't fully connected")
-
-        # We want to iterate over atoms that are good to use as reference
-        # points, but haven't been processed yet.
-        todo = ref_points.difference(processed)
-        for atom in todo:
-            for b in atom.bonds:
-                other = b.partner(atom)
-                # Avoid atoms not in our scope
-                if other not in ag_set:
-                    continue
-                if other in ref_points:
-                    continue
-                if b.length(pbc=False) > box_length:
-                    # Vector from ref atom to other
-                    vec = other.position - ref.position
-                    # Apply pbc to this vector
-                    vec -= np.rint(vec/box) * box
-                    # Define the position of other based on this vector
-                    other.position = ref.position + vec
-                # This atom can now be used as a reference point
-                ref_points.add(other)
-
-            processed.add(atom)
+        raise ValueError("Box lengths are too small relative to bond lengths")    
+    tric = triclinic_vectors(atomgroup.dimensions)
+    halfbox = tric.sum(axis=0) / 2.0
+    atgp = atomgroup
+    diffs = None
+    count = 0
+    idx = None
+    prev = {}
+    repeats = 0
+    while True:
+        # this method is iterative, as correcting a bond causes other to become distorted
+        if diffs is not None:
+            if idx is not None:
+                prev[count] = idx
+            idx = np.unique(np.where(diffs)[0])
+            atgp = atgp.bonds.atom2[idx]
+        # get the bond vectors and translate them to the center of the unit cell
+        # all the vectors that are bigger than half the unit cell will be outside
+        # these are the ones we want to "correct"
+        vecs = atgp.bonds.atom2.positions - atgp.bonds.atom1.positions
+        vecs += halfbox
+        # apply PBC on the vectors
+        newvecs = apply_PBC(vecs, atgp.dimensions)
+        # diffs is used to check which the atoms are already in their correct positions
+        # if all of them are correct then break the loop
+        diffs = np.any(~np.isclose(vecs, newvecs, 1e-6), axis=1)
+        if not diffs.sum():
+            break
+        # the new vectors are translated back to the origin and positions are updated
+        newvecs -= halfbox
+        atgp.bonds.atom2.positions = atgp.bonds.atom1.positions + newvecs
+        
+        # the following check is for cases when an atom in the atom2 group
+        # is connected to two atoms in atom1. Since where changing the positions
+        # of the atom2 group, suchs cases lead to the while loop being stuck:
+        # this atom is moved back and forth between the two atoms it is conected
+        # to. Since were tracking the atoms that need changing, if this group stays
+        # the same over multiple iterations, we change atom1.positions instead.
```

Comment:
@davidercruz If you had a ring molecule which had oddly/pathologically numbered bonds, I think you could end up chasing your tail because you're updating all positions at once.  This would lead to atoms from the ring jumping between each other images.